File: source/RevitDevTool.Test/BoundingBoxVisualization.cs
using System.Diagnostics;
using Autodesk.Revit.Attributes;
using Autodesk.Revit.UI.Selection;
using Nice3point.Revit.Toolkit.External;

namespace RevitDevTool.Test;

[Transaction(TransactionMode.Manual)]
[UsedImplicitly]
public class BoundingBoxVisualization : ExternalCommand
{
    public override void Execute()
    {
        try
        {
            var elementRef = UiDocument.Selection.PickObject(ObjectType.Element, "Select Element");
            var element = Document.GetElement(elementRef);

            var bbox = element.get_BoundingBox(ActiveView);
            Trace.Write(bbox);
        }
        catch (Autodesk.Revit.Exceptions.OperationCanceledException)
        {
           //
        }
        catch (Exception ex)
        {
            Trace.TraceError($"Error in BoundingBoxVisualization: {ex.Message}");
        }
    }
}

[Transaction(TransactionMode.Manual)]
[UsedImplicitly]
public class BoundingBoxesVisualization : ExternalCommand
{
    public override void Execute()
    {
        try
        {
            var elementRefs = UiDocument.Selection.PickObjects(ObjectType.Element, "Select Elements");
            var boxes = new List<BoundingBoxXYZ>();
            foreach (var elementRef in elementRefs)
            {
                var element = Document.GetElement(elementRef);
                var bbox = element.get_BoundingBox(ActiveView);
                if (bbox != null)
                {
                    boxes.Add(bbox);
                }
                else
                {
                    Trace.TraceWarning($"Element {element.Id} has no bounding box.");
                }
            }

            if (boxes.Count == 0)
            {
                Trace.TraceWarning("No bounding boxes found.");
                return;
            }

            Trace.Write(boxes);
        }
        catch (Exception ex)
        {
            Trace.TraceError($"Error in BoundingBoxesVisualization: {ex.Message}");
        }
    }
}

[Transaction(TransactionMode.Manual)]
[UsedImplicitly]
public class OutlineVisualization : ExternalCommand
{
    public override void Execute()
    {
        try
        {
            var elementRef = UiDocument.Selection.PickObject(ObjectType.Element, "Select Element");
            var element = Document.GetElement(elementRef);

            var bbox = element.get_BoundingBox(ActiveView);
            var outline = new Outline(bbox.Min, bbox.Max);
            outline.Scale(2);
            Trace.Write(outline);
        }
        catch (Autodesk.Revit.Exceptions.OperationCanceledException)
        {
            //
        }
        catch (Exception ex)
        {
            Trace.TraceError($"Error in BoundingBoxVisualization: {ex.Message}");
        }
    }
}

[Transaction(TransactionMode.Manual)]
[UsedImplicitly]
public class OutlinesVisualization : ExternalCommand
{
    public override void Execute()
    {
        try
        {
            var elementRefs = UiDocument.Selection.PickObjects(ObjectType.Element, "Select Elements");
            var boxes = new List<Outline>();
            foreach (var elementRef in elementRefs)
            {
                var element = Document.GetElement(elementRef);
                var bbox = element.get_BoundingBox(ActiveView);
                if (bbox != null)
                {
                    var ol = new Outline(bbox.Min, bbox.Max);
                    ol.Scale(2);
                    boxes.Add(ol);
                }
                else
                {
                    Trace.TraceWarning($"Element {element.Id} has no bounding box.");
                }
            }

            if (boxes.Count == 0)
            {
                Trace.TraceWarning("No bounding boxes found.");
                return;
            }

            Trace.Write(boxes);
        }
        catch (Exception ex)
        {
            Trace.TraceError($"Error in BoundingBoxesVisualization: {ex.Message}");
        }
    }
}

File: Samples/FSharpDemo/BoundingBoxVisualization.fs
namespace FSharpDemo

open System.Diagnostics
open Autodesk.Revit.Attributes
open Autodesk.Revit.DB
open Autodesk.Revit.UI.Selection
open JetBrains.Annotations
open Nice3point.Revit.Toolkit.External

[<Transaction(TransactionMode.Manual)>]
[<UsedImplicitly>]
type BoundingBoxVisualization() =
    inherit ExternalCommand()

    override this.Execute() =
        try
            let elementRef = this.UiDocument.Selection.PickObject(ObjectType.Element, "Select Element")
            let element = this.Document.GetElement(elementRef)
            Trace.Write(elementRef.ElementId)
            Trace.Write(element.get_Parameter(BuiltInParameter.IFC_GUID) |> Option.ofObj |> Option.map (fun p -> p.AsString()) |> Option.defaultValue null)
            Trace.Write(element.UniqueId)
            let bbox = element.get_BoundingBox(this.ActiveView)
            Trace.Write(bbox)
        with
        | :? Autodesk.Revit.Exceptions.OperationCanceledException -> ()
        | ex -> Trace.TraceError($"Error in BoundingBoxVisualization: {ex.Message}")

[<Transaction(TransactionMode.Manual)>]
[<UsedImplicitly>]
type BoundingBoxesVisualization() =
    inherit ExternalCommand()

    override this.Execute() =
        try
            let elementRefs = this.UiDocument.Selection.PickObjects(ObjectType.Element, "Select Elements")
            let boxes =
                [ for elementRef in elementRefs do
                    let element = this.Document.GetElement(elementRef)
                    let bbox = element.get_BoundingBox(this.ActiveView)
                    if bbox <> null then
                        yield bbox
                    else
                        Trace.TraceWarning($"Element {element.Id} has no bounding box.") ]

            if boxes.IsEmpty then
                Trace.TraceWarning("No bounding boxes found.")
            else
                Trace.Write(boxes)
        with
        | ex -> Trace.TraceError($"Error in BoundingBoxesVisualization: {ex.Message}")

[<Transaction(TransactionMode.Manual)>]
[<UsedImplicitly>]
type OutlineVisualization() =
    inherit ExternalCommand()

    override this.Execute() =
        try
            let elementRef = this.UiDocument.Selection.PickObject(ObjectType.Element, "Select Element")
            let element = this.Document.GetElement(elementRef)
            let bbox = element.get_BoundingBox(this.ActiveView)
            let outline = new Outline(bbox.Min, bbox.Max)
            outline.Scale(2.0)
            Trace.Write(outline)
        with
        | :? Autodesk.Revit.Exceptions.OperationCanceledException -> ()
        | ex -> Trace.TraceError($"Error in BoundingBoxVisualization: {ex.Message}")

[<Transaction(TransactionMode.Manual)>]
[<UsedImplicitly>]
type OutlinesVisualization() =
    inherit ExternalCommand()

    override this.Execute() =
        try
            let elementRefs = this.UiDocument.Selection.PickObjects(ObjectType.Element, "Select Elements")
            let boxes =
                [ for elementRef in elementRefs do
                    let element = this.Document.GetElement(elementRef)
                    let bbox = element.get_BoundingBox(this.ActiveView)
                    if bbox <> null then
                        let ol = new Outline(bbox.Min, bbox.Max)
                        ol.Scale(2.0)
                        yield ol
                    else
                        Trace.TraceWarning($"Element {element.Id} has no bounding box.") ]

            if boxes.IsEmpty then
                Trace.TraceWarning("No bounding boxes found.")
            else
                Trace.Write(boxes)
        with
        | ex -> Trace.TraceError($"Error in BoundingBoxesVisualization: {ex.Message}")

